src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -14948,6 +14947,66 @@ void M()
             c.VerifyDiagnostics();
         }
 
+        [Fact]
+        [WorkItem(29295, "https://github.com/dotnet/roslyn/issues/29295")]
+        public void CatchException_NullableType()
+        {
+            var c = CreateCompilation(@"
+class C
+{
+    void M()
+    {
+        try
+        {
+            System.Console.WriteLine();
+        }
+        catch (System.Exception? e)
+        {
+            var e2 = Copy(e);
+            e2.ToString();
+            e2 = null; // 1
+            e.ToString();
+        }
+    }
+    static U Copy<U>(U u) => throw null!;
+}
+", options: WithNonNullTypesTrue());
+
+            // There is still an issue with catch variables having an "unkonwn" null-state, instead of being initialized with a "not null" state
+            // We expect a warning on // 1
+            // Tracked by https://github.com/dotnet/roslyn/issues/33540
+            c.VerifyDiagnostics();
+        }
+
+        [Fact]
+        [WorkItem(29295, "https://github.com/dotnet/roslyn/issues/33540")]
+        public void CatchException_ConstrainedGenericTypeParameter()
+        {
+            var c = CreateCompilation(@"
+class C
+{
+    static void M<T>() where T : System.Exception?
+    {
+        try
+        {
+        }
+        catch (T e)
+        {
+            var e2 = Copy(e);
+            e2.ToString(); // 1
```

Comment:
Shouldn't `e` and `e2` both be `T!`, so no warning on `// 1`?

Reply:
`T!` is not a speakable type. The `var` should just be a `T`. But it is a `T~` (which is speakable with use of `#nullable disable`).
I'll add `e2 = default` to further illustrate the issue.